File: src/commonTest/kotlin/com/scoreo/application/UpdateMatchUseCaseTest.kt
package com.scoreo.application

import com.scoreo.infrastructure.InMemoryMatchRepository
import com.scoreo.domain.model.Match
import com.scoreo.domain.model.PlayerScore
import kotlin.test.Test
import kotlin.test.assertEquals

class UpdateMatchUseCaseTest {
    private lateinit var matchRepository: InMemoryMatchRepository
    private lateinit var useCase: UpdateMatchUseCase

    private fun setup() {
        matchRepository = InMemoryMatchRepository()
        useCase = UpdateMatchUseCase(matchRepository)
    }

    @Test
    fun testUpdateExistingMatch() {
        setup()
        val original = Match(
            id = "m1",
            gameTypeId = "g1",
            date = 1000L,
            playerScores = listOf(PlayerScore("p1", 10))
        )
        matchRepository.save(original)

        val updated = original.copy(playerScores = listOf(PlayerScore("p1", 20)))
        useCase.invoke(updated)

        val loaded = matchRepository.findById("m1")
        assertEquals(20, loaded?.playerScores?.get(0)?.score)
    }

    @Test
    fun testUpdatePreservesMatchId() {
        setup()
        val original = Match(
            id = "m1",
            gameTypeId = "g1",
            date = 1000L,
            playerScores = listOf()
        )
        matchRepository.save(original)

        val updated = original.copy(playerScores = listOf(PlayerScore("p1", 5)))
        useCase.invoke(updated)

        assertEquals("m1", matchRepository.getAll()[0].id)
    }

    @Test
    fun testUpdatePreservesDate() {
        setup()
        val original = Match(
            id = "m1",
            gameTypeId = "g1",
            date = 1000L,
            playerScores = listOf(PlayerScore("p1", 10))
        )
        matchRepository.save(original)

        val updated = original.copy(playerScores = listOf(PlayerScore("p1", 20)))
        useCase.invoke(updated)

        val loaded = matchRepository.findById("m1")
        assertEquals(1000L, loaded?.date)
    }
}

File: apps/scoreo/src/application/updateMatchUseCase.test.ts
import { describe, expect, it } from 'vitest'
import { InMemoryMatchRepository } from '../infrastructure/testing/inMemoryMatchRepository'
import { UpdateMatchUseCase } from './updateMatchUseCase'

describe('UpdateMatchUseCase', () => {
  it('updates an existing match', () => {
    const repo = new InMemoryMatchRepository()
    const useCase = new UpdateMatchUseCase(repo)
    const original = {
      id: 'm1',
      gameTypeId: 'g1',
      date: 1000,
      playerScores: [{ playerId: 'p1', score: 10 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    }
    repo.save(original)

    useCase.invoke({ ...original, playerScores: [{ playerId: 'p1', score: 20 }] })

    expect(repo.findById('m1')?.playerScores[0].score).toBe(20)
  })

  it('preserves match id', () => {
    const repo = new InMemoryMatchRepository()
    const useCase = new UpdateMatchUseCase(repo)
    const original = {
      id: 'm1',
      gameTypeId: 'g1',
      date: 1000,
      playerScores: [],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    }
    repo.save(original)

    useCase.invoke({ ...original, playerScores: [{ playerId: 'p1', score: 5 }] })

    expect(repo.getAll()[0].id).toBe('m1')
  })

  it('preserves date', () => {
    const repo = new InMemoryMatchRepository()
    const useCase = new UpdateMatchUseCase(repo)
    const original = {
      id: 'm1',
      gameTypeId: 'g1',
      date: 1000,
      playerScores: [{ playerId: 'p1', score: 10 }],
      manualWinners: [],
      secondaryPlayerScores: [],
      rounds: [],
    }
    repo.save(original)

    useCase.invoke({ ...original, playerScores: [{ playerId: 'p1', score: 20 }] })

    expect(repo.findById('m1')?.date).toBe(1000)
  })
})
